apis/core.oam.dev/v1beta1/resourcetracker_types.go

```diff
@@ -0,0 +1,274 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"bytes"
+	"encoding/json"
+	"strings"
+
+	errors2 "github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/common"
+	"github.com/oam-dev/kubevela/apis/interfaces"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/utils/errors"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// An ResourceTracker represents a tracker for track cross namespace resources
+// +kubebuilder:resource:scope=Cluster,categories={oam},shortName=rt
+type ResourceTracker struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ResourceTrackerSpec   `json:"spec,omitempty"`
+	Status ResourceTrackerStatus `json:"status,omitempty"`
+}
+
+// ResourceTrackerType defines the type of resourceTracker
+type ResourceTrackerType string
+
+const (
+	// ResourceTrackerTypeRoot means resources in this resourceTracker will only be recycled when application is deleted
+	ResourceTrackerTypeRoot = ResourceTrackerType("root")
+	// ResourceTrackerTypeVersioned means resources in this resourceTracker will be recycled when this version is unused and this resource is not managed by latest RT
+	ResourceTrackerTypeVersioned = ResourceTrackerType("versioned")
+	// ResourceTrackerTypeComponentRevision stores all component revisions used
+	ResourceTrackerTypeComponentRevision = ResourceTrackerType("component-revision")
+)
+
+// ResourceTrackerSpec define the spec of resourceTracker
+type ResourceTrackerSpec struct {
+	Type                  ResourceTrackerType `json:"type,omitempty"`
+	ApplicationGeneration int64               `json:"applicationGeneration"`
+	ManagedResources      []ManagedResource   `json:"managedResources,omitempty"`
+}
+
+// ManagedResource define the resource to be managed by ResourceTracker
+type ManagedResource struct {
+	common.ClusterObjectReference `json:",inline"`
+	common.OAMObjectReference     `json:",inline"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Data *runtime.RawExtension `json:"raw,omitempty"`
+	// Deleted marks the resource to be deleted
+	Deleted bool `json:"deleted,omitempty"`
+}
+
+// Equal check if two managed resource equals
+func (in ManagedResource) Equal(r ManagedResource) bool {
+	if !in.ClusterObjectReference.Equal(r.ClusterObjectReference) {
+		return false
+	}
+	if !in.OAMObjectReference.Equal(r.OAMObjectReference) {
+		return false
+	}
+	if in.Data != nil && r.Data != nil {
+		return bytes.Equal(in.Data.Raw, r.Data.Raw)
+	}
+	return (in.Data != nil) == (r.Data != nil)
```

Comment:
do you mean you just want to check the Data is all nil or non-nil? Do you want to check their data equal?

Reply:
Yes, need to check their data equal. The function here is not in-use now. Shall we remove it?

Reply:
Of course you should remove them. Dead code can cause confusion

Reply:
refactored to `reflect.DeepEqual`